### ISSUE TITLE
The alpha issue (premultiplicaiton and overloading)

### ISSUE THREAD

**Issue by ritchse** [ojack/hydra-synth#109]
# The alpha issue (premultiplicaiton and overloading)

I create this issue after discussion in the Discord and various hours of testing different ideas. Thanks to @BenjaminHolland for the resources he shared in #107 which have been very helpful on finding weird behaviours while modifying alpha.

---

### **What is premultiplication?**

```
If an alpha channel is used in an image, there are two common representations that are available: straight (unassociated) alpha and premultiplied (associated) alpha.

* With straight alpha, the RGB components represent the color of the object or pixel, disregarding its opacity.
* With premultiplied alpha, the RGB components represent the emission of the object or pixel, and the alpha represents the occlusion. 
```
[source](https://en.wikipedia.org/wiki/Alpha_compositing#Straight_versus_premultiplied)

Basically:

```glsl
//straight alpha
return vec4(color, alpha)
//premultiflied alpha
return vec4(color*alpha, alpha)
```

Alpha compositing and different operations must be done consistently with the type of alpha being used.
Unfortunately, Hydra is inconsistent with the way of handling alpha.

### **The inconsistency**

Functions that output / assume and input of premultiplied alpha:
* luma (outputs)
* mask (outputs)
* layer (inputs, almost)
* * The current implementation is weird, it doesn't work correctly with neither actually. And also has [this bug](https://hydra.ojack.xyz/?sketch_id=MCjFkV0bE27KW4Li)

Functions that output / input straight alpha:
* Literally the whole rest of them that use color

This inconsistency can be seen in the following example, taken from @micuat 's [2nd tutorial on feedback](https://youtu.be/U5Li6n_zKlE?t=517).

```js
src(o0)
	.modulateHue(src(o0).scale(1.01),1)
	.layer(
  		osc(10,.15,2)
  			.mask(
             	shape(4,0.5,.1)
            )
  	)
	.out(o0)
```
[open in editor](https://hydra.ojack.xyz/?code=YXdhaXQlMjBsb2FkU2NyaXB0KCdodHRwcyUzQSUyRiUyRmh5ZHJhLWV4dGVuc2lvbnMuZ2xpdGNoLm1lJTJGaHlkcmEtYWxwaGEtaXNzdWUtcGxheWdyb3VuZC5qcycpJTBBJTBBc3JjKG8wKSUwQSUwOS5tb2R1bGF0ZUh1ZShzcmMobzApLnNjYWxlKDEuMDEpJTJDMSklMEElMDkubGF5ZXIoJTBBJTIwJTIwJTA5JTA5b3NjKDEwJTJDLjE1JTJDMiklMEElMjAlMjAlMDklMDklMDkubWFzayglMEElMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMDlzaGFwZSg0JTJDMC41JTJDLjEpJTBBJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwKSUwQSUyMCUyMCUwOSklMEElMDkub3V0KG8wKQ==)

See how using pre-multiplied (abbreviated as pm) alpha versions of `layer` and `mask` gives the expected result. Actually, this is using the current version of `mask`.

```js
await loadScript('https://hydra-extensions.glitch.me/hydra-alpha-issue-playground.js')

src(o0)
	.modulateHue(src(o0).scale(1.01),1)
	.pmlayer(
  		osc(10,.15,2)
  			.pmmask(
             	shape(4,0.5,.1)
            )
  	)
	.out(o0)
```
[open in editor](https://hydra.ojack.xyz/?code=YXdhaXQlMjBsb2FkU2NyaXB0KCdodHRwcyUzQSUyRiUyRmh5ZHJhLWV4dGVuc2lvbnMuZ2xpdGNoLm1lJTJGaHlkcmEtYWxwaGEtaXNzdWUtcGxheWdyb3VuZC5qcycpJTBBJTBBc3JjKG8wKSUwQSUwOS5tb2R1bGF0ZUh1ZShzcmMobzApLnNjYWxlKDEuMDEpJTJDMSklMEElMDkucG1sYXllciglMEElMjAlMjAlMDklMDlvc2MoMTAlMkMuMTUlMkMyKSUwQSUyMCUyMCUwOSUwOSUwOS5wbW1hc2soJTBBJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTA5c2hhcGUoNCUyQzAuNSUyQy4xKSUwQSUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCklMEElMjAlMjAlMDkpJTBBJTA5Lm91dChvMCk=)

```glsl
vec4 pmlayer(vec4 _c0, vec4 _c1) {
    return _c1+(_c0*(1.0-_c1.a));
}

vec4 pmmask(vec4 _c0, vec4 _c1) {
    float a = _luminance(_c1.rgb);
    return vec4(_c0.rgb*a, _c0.a*a);      
}
```
And now with straight (st) alpha:

```js
await loadScript('https://hydra-extensions.glitch.me/hydra-alpha-issue-playground.js')

src(o0)
	.modulateHue(src(o0).scale(1.01),1)
	.stlayer(
  		osc(10,.15,2)
  			.stmask(
             	shape(4,0.5,.1)
            )
  	)
	.out(o0)
```
[open in editor](https://hydra.ojack.xyz/?code=YXdhaXQlMjBsb2FkU2NyaXB0KCdodHRwcyUzQSUyRiUyRmh5ZHJhLWV4dGVuc2lvbnMuZ2xpdGNoLm1lJTJGaHlkcmEtYWxwaGEtaXNzdWUtcGxheWdyb3VuZC5qcycpJTBBJTBBc3JjKG8wKSUwQSUwOS5tb2R1bGF0ZUh1ZShzcmMobzApLnNjYWxlKDEuMDEpJTJDMSklMEElMDkuc3RsYXllciglMEElMjAlMjAlMDklMDlvc2MoMTAlMkMuMTUlMkMyKSUwQSUyMCUyMCUwOSUwOSUwOS5zdG1hc2soJTBBJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTA5c2hhcGUoNCUyQzAuNSUyQy4xKSUwQSUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCklMEElMjAlMjAlMDkpJTBBJTA5Lm91dChvMCk=)

```glsl
vec4 stlayer(vec4 _c0, vec4 _c1) {
    float a = _c1.a + (_c0.a*(1.0-_c1.a));
    vec3 color = (_c1.rgb*_c1.a) + 
                    (_c0.rgb*_c0.a*(1.0-_c1.a));
    color /= a + 0.0000001;
    return vec4(color,a);
}

vec4 stmask(vec4 _c0, vec4 _c1) {
    float a = _luminance(_c1.rgb);
    return vec4(_c0.rgb, _c0.a*a);   
}
```

Both layering methods were taking from [the alpha composition wikipedia page](https://en.wikipedia.org/wiki/Alpha_compositing)

### **The screen lies to you**

```js
osc().color(1,1,1,0)
	.out()
```

There should be nothing in the screen, yet the bright colors are still showing. (If you change the website's background to red you'll see the oscillator with a red background instead of black). I think this is because by default WebGL assumes you'll use premultiplied alpha.

See [this stackoverflow question](https://stackoverflow.com/questions/20362023/webgl-why-does-transparent-canvas-show-clearcolor-color-component-when-alpha-is).

### **Straight vs premultiplied, what should Hydra use internally?**

I believe Hydra should use straight alpha internally (at least by default), most of the functions already work that way and I believe that's the default mindset of any shaderer. Also because premultiplication is a lossy transformation (you can't get the exact same rgb colors back after premultiplying). In the long run we could build an option into Hydra to set which alpha mode to use, since there has also been talk about enabling/disabling wrapping. Also, PNG images use straight alpha, so setting the default to be premultiplied would also imply taking care of that. This is also why PNG images look so weird in Hydra unless you layer them:

```js
await loadScript("https://hydra-extensions.glitch.me/hydra-glsl.js")

s0.initImage('https://i.imgur.com/3KPzyjx.png')
src(s0)
	//.glslColor('float a = clamp(c0.a,0.0,1.0); vec4(c0.rgb*a,a)') //premultiplication
	.out(o0)
```

### **Proposed solution:**

* Set every function to work with straight alpha
* Premultiply the final output set in [`renderFbo()`](https://github.com/ojack/hydra-synth/blob/fe5414dd889e4eed61bddc10e9649cb1fd622573/hydra-synth.js#L312) and [`renderAll()`](https://github.com/ojack/hydra-synth/blob/fe5414dd889e4eed61bddc10e9649cb1fd622573/hydra-synth.js#L257)

Output premultiplication could be as such:

```glsl
void main () {
    [...]
    float alpha = clamp(gl_FragColor.a, 0.0, 1.0);
    gl_FragColor.rgb *= alpha;
    gl_FragColor.a = alpha;
}
```

#### **Drawbacks**

* The changes in `mask` will have a visual impact, but this is technically a fix, as we see in the example above.
* The changes in `luma` will be somewhat noticeable too (at least when layering over opaque backgrounds). `luma` currently leaves some strong dark edges (caused by the premultiplication) which various users already expect from `luma`. Correcting luma will lighten them a bit.

These also have to do with the correction of `layer`.

##### **`luma` example**:

```js
await loadScript('https://hydra-extensions.glitch.me/hydra-alpha-issue-playground.js')

solid(1,.5,0)
	.stlayer(osc().stluma(.5,.1))
	.out(o0)
solid(1,.5,0)
	.layer(osc().luma(.5,.1))
	.out(o1)
solid(0,.5,1)
	.stlayer(osc(60,.1,2).stluma())
	.out(o2)
solid(0,.5,1)
	.layer(osc(60,.1,2).luma())
	.out(o3)

render()
```

I think the changes are light enough for it not to be too big of a deal. But we can vote on the Discord or here maybe.

## On overloading of alpha values

Using Hydra is very common to overload alpha values while using `add` or `color`. Or cancelling them out with `sub` (and not realizing because currently bright colors with 0 alpha still appear on the screen!).

#### `add`

Adding two images in Hydra also adds their alpha. In most cases this ends up with an image full of alpha values of 2.0.
This isn't much of a problem if we set on a specific way of handling alpha. For example, the straight alpha version of layer handles overloaded alpha with no problem:

```js
await loadScript('https://hydra-extensions.glitch.me/hydra-alpha-issue-playground.js')

b = ()=> osc().add(noise())
b().out()
solid(0,0,0,0).stlayer(b()).out(o1)
src(o1).sub(o0).out(o2)
b().sub(solid(0,0,0,0).stlayer(b())).out(o3)

// change st layer to layer and see the original problem

render()
```
[open in editor](https://hydra.ojack.xyz/?code=YXdhaXQlMjBsb2FkU2NyaXB0KCdodHRwcyUzQSUyRiUyRmh5ZHJhLWV4dGVuc2lvbnMuZ2xpdGNoLm1lJTJGaHlkcmEtYWxwaGEtaXNzdWUtcGxheWdyb3VuZC5qcycpJTBBJTBBYiUyMCUzRCUyMCgpJTNEJTNFJTIwb3NjKCkuYWRkKG5vaXNlKCkpJTBBYigpLm91dCgpJTBBc29saWQoMCUyQzAlMkMwJTJDMCkuc3RsYXllcihiKCkpLm91dChvMSklMEFzcmMobzEpLnN1YihvMCkub3V0KG8yKSUwQWIoKS5zdWIoc29saWQoMCUyQzAlMkMwJTJDMCkuc3RsYXllcihiKCkpKS5vdXQobzMpJTBBJTBBcmVuZGVyKCklMEElMEElMEElMEElMEE=)

Therefore in this scenario most users won't notice the overloading. No current Hydra function will suffer from this. However almost all of the functions proposed in #107 will. This could be fixed by clamping alpha before these specific functions. However the biggest intuition problem comes with alpha cancellation:

#### `sub`

Substracting an opaque image from another results in a blank image (and again, the user can't realize because bright colors still appear currently even if they have 0 alpha). This can be very couter-intuitive and the current way of leading with this would be to add `.color(1,1,1,0)` to the input of `sub`, which is tedious. My first idea was to change both addition and substraction to:

```glsl
void add(vec4 _c0, vec4 _c1, float amount, float alpha){
    _c1.a *= alpha;
    return _c0+(_c1*amount);
}
void sub(vec4 _c0, vec4 _c1, float amount, float alpha){
    _c1.a *= alpha;
    return _c0+(_c1*amount);
}
```

However this would mean that when `alpha` is set to 0, when adding, it wouldn't overload, but it also wouldn't add alpha at all. Which is counter-intuitive if the image to which the `add` function is being applied to has transparencies. It would be like saying `a().add(b().mask(a()))`.

A solution could be to treat alpha in the same way that `diff` does:

```glsl
void add(vec4 _c0, vec4 _c1, float amount){
    return vec4(_c0.rgb+(_c1.rgb*amount),max(_c0.a,_c1.a));
}
void sub(vec4 _c0, vec4 _c1, float amount){
    return vec4(_c0.rgb+(_c1.rgb*amount),max(_c0.a,_c1.a));
}
```

Return the maximum alpha value from each input. This could be a great solution, since it's super rare to want to use overloaded values from alpha. And these rare cases could be achived by doing `.color(1,1,1,2) // or any n > 1`, which is more intuitive than having to do `.color(1,1,1,0)` everytime images are substracted.

## Other ideas

* Adding a `normalize` or `clamp` functions that simply clamps every channel to be between 0 and 1, could be useful for some arithmetics.
* Actually add the functions in #107
* * Specially the `source over`, which I'd call `unlayer`. Basically layers the GlslSource to which it is being applied over the one provided in the argument.
* Add a `div` function for more arithmetics.

**Comment by ritchse** [comment]
The `noise` function would also need to be changed a bit. Currently it outputs with values in the range [-1,1]. This is what causes [this bug](https://hydra.ojack.xyz/?code=YXdhaXQlMjBsb2FkU2NyaXB0KCdodHRwcyUzQSUyRiUyRmh5ZHJhLWV4dGVuc2lvbnMuZ2xpdGNoLm1lJTJGaHlkcmEtZ2xzbC5qcycpJTBBJTBBc29saWQoMSUyQzAlMkMwKSUwQSUwOS5sYXllciglMEElMjAlMjAlMDklMDlvc2MoOCUyQy4xJTJDMiklMEElMjAlMjAlMDklMDklMDkubWFzayhzaGFwZSgpKSUwQSUwOSklMEElMDkub3V0KCklMEElMEFzb2xpZCgxJTJDMCUyQzApJTBBJTA5LmxheWVyKCUwQSUyMCUyMCUwOSUwOW9zYyg4JTJDLjElMkMyKSUwQSUyMCUyMCUwOSUwOSUwOS5tYXNrKG5vaXNlKDQpKSUyMCUyRiUyRiUyMG5lZ2F0aXZlJTIwdmFsdWVzJTIwaW4lMjBtYXNrJTIwYnJlYWslMjBpdCUyMCElMEElMDkpJTBBJTA5Lm91dChvMSklMEElMEFzb2xpZCgxJTJDMCUyQzApJTBBJTA5LmxheWVyKCUwQSUyMCUyMCUwOSUwOW9zYyg4JTJDLjElMkMyKSUwQSUyMCUyMCUwOSUwOSUwOS5tYXNrKG5vaXNlKDQpLmdsc2xDb2xvcignbWF4KGMwJTJDMC4wKScpKSUwQSUwOSklMEElMDkub3V0KG8yKSUwQSUwQXNvbGlkKDElMkMwJTJDMCklMEElMDkubGF5ZXIoJTBBJTIwJTIwJTA5JTA5b3NjKDglMkMuMSUyQzIpJTBBJTIwJTIwJTA5JTA5JTA5Lm1hc2sobm9pc2UoNCkudGhyZXNoKCkpJTBBJTA5KSUwQSUwOS5vdXQobzMpJTBBJTBBcmVuZGVyKCk=).

**[Noise modifications](https://hydra.ojack.xyz/?code=YXdhaXQlMjBsb2FkU2NyaXB0KCdodHRwcyUzQSUyRiUyRmh5ZHJhLWV4dGVuc2lvbnMuZ2xpdGNoLm1lJTJGaHlkcmEtZ2xzbC5qcycpJTBBJTBBbm9pc2UoNSUyQzEpJTBBJTA5Lm91dCgpJTBBJTBBbm9pc2UoNSUyQzEpJTBBJTA5Lmdsc2xDb2xvcignMC41JTJCKGMwJTJGMi4pJyklMjAlMkYlMkYlMjB0aGUlMjBhY3R1YWwlMjBub2lzZSUyQyUyMGJ1dCUyMGNoYW5nZXMlMjB0b28lMjBtdWNoJTIwZnJvbSUyMHRoZSUyMG9yaWdpbmFsJTBBJTA5Lm91dChvMSklMEElMEFub2lzZSg1JTJDMSklMEElMDkuZ2xzbENvbG9yKCdtYXgoYzAlMkMwLjApJyklMjAlMkYlMkYlMjBzbG93JTIwcGVyZm9ybWFuY2Utd2lzZSUyQyUyMGJ1dCUyMGFjY2VwdGFibGUlMjBpbW8lMkMlMjBhbHRob3VnaCUyMGl0JTIwaXMlMjBsb3NzeSUwQSUwOS5vdXQobzIpJTA5JTA5JTA5JTA5JTA5JTJGJTJGJTIwd2UlMjBjb3VsZCUyMGRvJTIwdGhpcyUyMGluc2lkZSUyMG9mJTIwJ21hc2snJTIwdGhvJTNCJTIwaSUyMGRvbid0JTIwc2VlJTIwbWFueSUyMHBlb3BsZSUyMG1ha2luZyUyMHVzZSUyMG9mJTIwYnJlYWtpbmclMjBtYXNrJTBBJTBBcmVuZGVyKCk=)**

**Comment by micuat** [comment]
I think it's important to think about "specify hydra-synth version" feature request because any of these will affect the current sketches
https://github.com/ojack/hydra/issues/113

and I'm so used to the way how alpha is treated in the current implementation for example

```javascript
src(o0)
	.modulateHue(src(o0).scale(1.01),1)
	.layer(
  		osc(10,.15,2).color(1,1,1,0)
  			.add(
             	shape(4,0.5,.1).r().color(0,0,0,1)
            )
  	)
	.out(o0)
```
https://hydra.ojack.xyz/?code=c3JjKG8wKSUwQSUwOS5tb2R1bGF0ZUh1ZShzcmMobzApLnNjYWxlKDEuMDEpJTJDMSklMEElMDkubGF5ZXIoJTBBJTIwJTIwJTA5JTA5b3NjKDEwJTJDLjE1JTJDMikuY29sb3IoMSUyQzElMkMxJTJDMCklMEElMjAlMjAlMDklMDklMDkuYWRkKCUwQSUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUwOXNoYXBlKDQlMkMwLjUlMkMuMSkucigpLmNvbG9yKDAlMkMwJTJDMCUyQzEpJTBBJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwKSUwQSUyMCUyMCUwOSklMEElMDkub3V0KG8wKQ==

**Comment by ritchse** [comment]
100% agree !

**Comment by ritchse** [comment]
btw your example would still look the same with the proposed changes!

[on editor](https://hydra.ojack.xyz/?code=JTBBYXdhaXQlMjBsb2FkU2NyaXB0KCdodHRwcyUzQSUyRiUyRmh5ZHJhLWV4dGVuc2lvbnMuZ2xpdGNoLm1lJTJGaHlkcmEtYWxwaGEtaXNzdWUtcGxheWdyb3VuZC5qcycpJTBBYXdhaXQlMjBsb2FkU2NyaXB0KCdodHRwcyUzQSUyRiUyRmh5ZHJhLWV4dGVuc2lvbnMuZ2xpdGNoLm1lJTJGaHlkcmEtZ2xzbC5qcycpJTBBc3JjKG8wKSUwQSUyMCUyMCUyMCUyMC5tb2R1bGF0ZUh1ZShzcmMobzApLnNjYWxlKDEuMDEpJTJDMSklMEElMjAlMjAlMjAlMjAuc3RsYXllciglMEElMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBvc2MoMTAlMkMuMTUlMkMyKS5jb2xvcigxJTJDMSUyQzElMkMwKSUwQSUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMC5hZGQoJTBBJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwc2hhcGUoNCUyQzAuNSUyQy4xKS5yKCkuY29sb3IoMCUyQzAlMkMwJTJDMSklMEElMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjApJTBBJTIwJTIwJTIwJTIwJTIwJTIwKSUwQSUyMCUyMCUyMCUyMC5vdXQobzApJTBBc3JjKG8wKSUwQSUyMCUyMCUyMCUyMC5nbHNsQ29sb3IoJ2Zsb2F0JTIwYSUyMCUzRCUyMGNsYW1wKGMwLmElMkMwLjAlMkMxLjApJTNCJTIwdmVjNChjMC5yZ2IqYSUyQ2EpJyklMEElMjAlMjAlMjAlMjAub3V0KG8xKSUwQXJlbmRlcihvMSk=)

the biggest change visually with the current proposals would be anything were the first generator in the sketch has no alpha, which currently still shows up, but these are very very rare cases.
the biggest change in terms of usage would be to `mask` and `luma` as mentioned.

btw are sketches saved with timestamps? i can't remember, will check tho. but that would make it easy to check the release date of each hydra version and just make a script that automatically assigns the correct version to every sketch already stored in the db. anyway, this is for the other issue

**Comment by ojack** [comment]
Thank you so much @ritchse for describing all of these in so much detail!

I agree about needing hydra version specification and to create versioned bundles of hydra-synth before integrating these changes. I use the mask "bug" a lot in order to create edges: https://hydra.ojack.xyz/?sketch_id=IlXoaQ34iJTaZn4X , even though I agree it is not expected behavior. 
<img width="732" alt="Screenshot 2022-02-08 at 12 49 34" src="https://user-images.githubusercontent.com/4909769/152981424-4be33fd7-8453-442b-a5e7-50423ac35165.png">

I wonder if there is a way to create edges with the new updates (or add that in as an extra parameter to mask and luma)?

**Comment by ojack** [comment]
> 

I am not sure yet that this is a bug! I kind of like having interesting behaviors for negative values of things. If anything, I wish more things would output negative color as it makes it really useful for feedback. It seems like a clamp function would solve this.

**Comment by micuat** [comment]
I second that this is *not a bug but a feature* or rather a *charm* :)

By the way, you might know but drawing to o0...o3 will clamp the output:

```javascript
render()

noise().out()
src(o0).out(o1)

noise().add(solid(1,1,0)).out(o2) // noise has negative values
src(o0).add(solid(1,1,0)).out(o3) // src(o0) has no negative values
```
https://hydra.ojack.xyz/?code=cmVuZGVyKCklMEElMEFub2lzZSgpLm91dCgpJTBBc3JjKG8wKS5vdXQobzEpJTBBJTBBbm9pc2UoKS5hZGQoc29saWQoMSUyQzElMkMwKSkub3V0KG8yKSUyMCUyRiUyRiUyMG5vaXNlJTIwaGFzJTIwbmVnYXRpdmUlMjB2YWx1ZXMlMEFzcmMobzApLmFkZChzb2xpZCgxJTJDMSUyQzApKS5vdXQobzMpJTIwJTJGJTJGJTIwc3JjKG8wKSUyMGhhcyUyMG5vJTIwbmVnYXRpdmUlMjB2YWx1ZXMlMEE=

if you want to clamp without rendering to a buffer, you can do something like

```javascript
noise().luma(0,0).invert().luma(0,0).invert()
```

**Comment by ritchse** [comment]
Sure! I was thinking mostly about the experience of doing `.mask(noise())` which is a bit unexpected. So I guess the question is more about clamping the input to `mask` rather. Or to just rely on a clamp function.

And yes, I know rendering to output clamps the image, but that's actually part of what made me concern in the first place. As in [this example from above](https://hydra.ojack.xyz/?code=YXdhaXQlMjBsb2FkU2NyaXB0KCdodHRwcyUzQSUyRiUyRmh5ZHJhLWV4dGVuc2lvbnMuZ2xpdGNoLm1lJTJGaHlkcmEtYWxwaGEtaXNzdWUtcGxheWdyb3VuZC5qcycpJTBBJTBBYiUyMCUzRCUyMCgpJTNEJTNFJTIwb3NjKCkuYWRkKG5vaXNlKCkpJTBBYigpLm91dCgpJTBBc29saWQoMCUyQzAlMkMwJTJDMCkuc3RsYXllcihiKCkpLm91dChvMSklMEFzcmMobzEpLnN1YihvMCkub3V0KG8yKSUwQWIoKS5zdWIoc29saWQoMCUyQzAlMkMwJTJDMCkuc3RsYXllcihiKCkpKS5vdXQobzMpJTBBJTBBcmVuZGVyKCklMEElMEElMEElMEElMEE=)

About the edges:
That's an interesting use case! An argument for mask and luma is doable. You could also `.mask(shape()).mult(shape())`, or something like`.layer(shape().luma(0.2,0.01))`.

**Comment by ojack** [comment]
I am finding that there are so many issues and solutions wrapped up into this one issue, so it is a bit hard to follow everything. This is my best at just defining the bugs and idiosyncracies that are mentioned here (without talking yet about the proposed solutions): 

1. Layering using soft / blurred edges causes dark edges. Affected functions: luma() and mask() . Example: https://hydra.ojack.xyz/?sketch_id=NAOIGIny5MRhmZ2Y

2. Arithmetic operations add() and sub() cause unexpected alpha changes. This behavior is inconsistent with other blending functions such as diff. Affected functions: add() and sub()

3. Mask and layer have strange behaviors when used with negative alpha values. Affected functions: mask() and layer() https://hydra.ojack.xyz/?sketch_id=0qi4VpzDiExxAobY

4. Layering with fully transparent layers still affects color of final output: https://hydra.ojack.xyz/?sketch_id=MCjFkV0bE27KW4Li
Affected functions: layer()

Am I missing anything?

**Comment by ojack** [comment]
@ritchse I was trying to get multiple composite operations to work with `stmask()` from above but couldn't figure it out.
As I recall this was why we switched to pre-multiplying at some point with `mask()` and `luma()`. I tried implementing `clamp()` and `add()` as I understood them from above. The top left is the current behavior, which is also what I would expect to happen. https://hydra.ojack.xyz/?sketch_id=hXWGlPAVfp3brgPX
![Screenshot 2022-02-08 at 17 37 22](https://user-images.githubusercontent.com/4909769/153032891-b87d9daf-a659-47d7-bb4f-48c79d1088e3.png)

**Comment by ritchse** [comment]
Yep, that's a good summary! Although mask actually has problems with negative color values, since it uses the _luminance function to generate the mask, which expects positive values.

That's technically right behavior because using straight alpha color values are left untouched. So when you add an osc that has a constant 1 alpha value, they come back. That's a point of intuitiveness for premultiplication! Technically what you're trying to do, since osc makes an opaque image, is `osc().mask(shape()).add(solid(1,1,1).mask(osc()))`.
But the solution you describe on the edit could work... But it does generate inconsistency, any addition will premult the output. 

Also, what you called clamp2 is actually premultiplication!
clamp would be `return clamp(_c0, 0.0, 1.0)`. When testing straight alpha you have to put some premultiplication at the end of it all. And if you have feedback on your sketch you have to `[...].out(o0); src(o0).premultiplication().out(o1)`.
I'll update the `hydra-alpha-issue-playground.js` and add clamp and premultiplication as functions when I have access to a computer.

**Comment by ritchse** [comment]
Thought about it a bit more. The bit of code I proposed for the mask issue won't work for the same reasons, mask doesn't affect color, so the image added would be a white solid. Gon have to think about it a bit more. I guess mayyyybe there's a way to mix st and pm in a rational way if the functions take care of it? I dunno, testing testing testing!

**Comment by ojack** [comment]
What I think could work is:
1) For operations that only affect alpha (`luma()` and `mask()`) do not multiply the rgb by alpha
2) For operations that combine rgb values (`layer()`,blend()`, `add()`, `diff()`) multiply the respective rgb values by their own alpha before calculating rgb. 

The above looks to me to be exactly what your `stlayer()` function (combined with `stmask()` and `stluma()` is doing already. i.e, performing the multiplication during the stlayer() rather than the stluma() (NOT waiting until the end of the chain because then you lose access to the alpha information from the two separate sources). Above was an attempt to do this with `add()` as well. 

I am deliberately not using the word premultiply right now because I think it got a bit muddled above....i.e. you always have to multiply the values at some point, so 'premultiply' only has meaning respective to whether it has happened *yet* for a given texture relative to a specific point in the chain of transformations. 

Need to test this a lot and probably has a lot of holes!!! Sometimes I really wish for a visual test script where you could see the output of lots of example scripts on separate small canvases.

**Comment by ritchse** [comment]
Well I come again to this issue as I'm working on some blending functions. @ojack I decided to now respond to your last comment because I feel I'll forget what I've just thought and I'll have to think about it again in the future if I don't write it down lol. But feel free to ignore this comment if you don't want to focus on this issue now.

I understand the term premultiplication it's not as correct here but it's more concise than saying "multiplying by alpha".

The reason stmask and stluma multiply c1.rgb by c1.a at the beginning is merely because the luminance of c1.rgb is being calculated. It's doing the multiplication there because retrieving luminance is a process that works with opaque textures. Multiplying by alpha here should be seen as converting from RGBA (straight alpha) to RGB so that the luminance can be calculated properly. If you use premultiplication all around, then logically, you have no need of multiplying c1.rgb by c1.a inside mask and luma, given you can assume c1 is a premultiplied source. What you are describing is basically setting on premultiplication. 

However I'm not sure what you mean by "not waiting until the end of the chain because then you lose access to the alpha information from the two separate sources". Maybe I'm not understanding what you mean, but multiplying by alpha is in itself a lossy process, leaving it until the end of the chain (regl's render shaders in hydra-synth.js) is the only way of not losing any information in any way in the middle of a patch. 

On the topic of those "edges" that you meantioned earlier about mask. That could totally be turned into a parameter for the function of stmask. It would basically be changing `c0.rgb` to `mix(c0.rgb, c0.rgb*a, amountOfEdges)`.

As mentioned above too, "fixing" these edges would break old sketches so of course there would have to be an effort for a versioning system (maybe for the hydra-editor v2?, i'm not sure how's that project going) before applying any changes.

Overall, taking into account what's been discussed in this issue, and just my experience over the last year using hydra and glsl while being aware of this issue, I'm pretty sure my vote goes for switching every function to work with straight alpha. Over the year I've loaded stlayer and stmask into my hydra sessions many many times just because they work "properly". I've also realized that regarding `noise`, the only problem with it is that its bipolarity isn't documented properly. I think `add` and `sub` should be left as is. Adding `alpha` as an extra argument to them, as a shortcut for `.add(tex.color(1,1,1,0))` could be a way to make it easier for users to avoid the problems of cancelling or overloading alpha. I don't think we should add the transforms in #107 anymore. I do think we should add division to Hydra though, merely for arithmetics, it feels weird all the basic operators are there but not division.
